### PR TITLE
Refactor `@trace_disabled` implementation

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/12285?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12285/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12285
```

</p>
</details>

### What changes are proposed in this pull request?

Follow-up for #12268 

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracing/provider.py [hunk 1]
```diff
@@ -244,15 +219,33 @@ def f():
...
     @functools.wraps(f)
     def wrapper(*args, **kwargs):
-        was_trace_enabled = _is_enabled()
-        if was_trace_enabled:
-            disable()
-            try:
+        is_func_called = False
+        try:
+            was_trace_enabled = _is_enabled()
+            if was_trace_enabled:
```
Comment (harupy) [inline mlflow/tracing/provider.py:225]:
```suggestion
            if _is_enabled():
```

nit

File: mlflow/tracing/provider.py [hunk 2]
```diff
@@ -244,15 +219,33 @@ def f():
...
+                    enable()
+            else:
+                is_func_called = True
+                return f(*args, **kwargs)
+        # We should only catch the exception from disable() and enable()
+        # and let other exceptions propagate.
+        except MlflowTracingException as e:
+            _logger.warning(
+                f"An error occurred while disabling or re-enabling tracing: {e} "
+                "The original function was still executed, but the tracing "
```
Comment (harupy) [inline mlflow/tracing/provider.py:240]: At this point, the original function may not be exeucted, right?

File: tests/tracing/utils/test_search.py
Comment (harupy) [inline tests/tracing/utils/test_search.py:1]: Thanks!

File: mlflow/tracing/provider.py [hunk 3]
```diff
@@ -244,15 +219,33 @@ def f():
...
+                finally:
+                    enable()
+            else:
+                is_func_called = True
+                return f(*args, **kwargs)
+        # We should only catch the exception from disable() and enable()
+        # and let other exceptions propagate.
+        except MlflowTracingException as e:
+            _logger.warning(
+                f"An error occurred while disabling or re-enabling tracing: {e} "
```
[the remark this answers is not in the document]
Comment (B-Step62) [reply mlflow/tracing/provider.py:239]: Ahh good catch, let me remove the prefix from `raise_as_trace_exception`

File: mlflow/tracing/provider.py (hunk 2 above)
Comment (B-Step62) [reply mlflow/tracing/provider.py:240]: True, more appropriate term should be `will be`

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**